src/ILCompiler.TypeSystem/tests/CastingTests.cs

```diff
@@ -0,0 +1,211 @@
+            TypeDesc uintType = _context.GetWellKnownType(WellKnownType.UInt32);
+            TypeDesc byteType = _context.GetWellKnownType(WellKnownType.Byte);
+            TypeDesc sbyteType = _context.GetWellKnownType(WellKnownType.SByte);
+            TypeDesc intPtrType = _context.GetWellKnownType(WellKnownType.IntPtr);
+            TypeDesc ulongType = _context.GetWellKnownType(WellKnownType.UInt64);
+
+            TypeDesc doubleType = _context.GetWellKnownType(WellKnownType.Double);
+            TypeDesc boolType = _context.GetWellKnownType(WellKnownType.Boolean);
+
+            TypeDesc intBasedEnumType = _testModule.GetType("Casting", "IntBasedEnum");
+            TypeDesc uintBasedEnumType = _testModule.GetType("Casting", "UIntBasedEnum");
+            TypeDesc shortBasedEnumType = _testModule.GetType("Casting", "ShortBasedEnum");
+
+            Assert.True(intType.MakeArrayType().CanCastTo(uintType.MakeArrayType()));
+            Assert.False(intType.CanCastTo(uintType));
+
+            Assert.True(byteType.MakeArrayType().CanCastTo(sbyteType.MakeArrayType()));
+            Assert.False(byteType.CanCastTo(sbyteType));
+
+            Assert.True(intPtrType.MakeArrayType().CanCastTo(ulongType.MakeArrayType()));
+            Assert.False(intPtrType.CanCastTo(ulongType));
+
+            // These are same size, but not allowed to cast
+            Assert.False(doubleType.MakeArrayType().CanCastTo(ulongType.MakeArrayType()));
+            Assert.False(boolType.MakeArrayType().CanCastTo(byteType.MakeArrayType()));
+
+            Assert.True(intBasedEnumType.MakeArrayType().CanCastTo(uintType.MakeArrayType()));
+            Assert.True(intBasedEnumType.MakeArrayType().CanCastTo(uintBasedEnumType.MakeArrayType()));
+            Assert.False(shortBasedEnumType.MakeArrayType().CanCastTo(intType.MakeArrayType()));
+        }
+
+        [Fact]
+        public void TestArrayInterfaceCasting()
+        {
+            TypeDesc intType = _context.GetWellKnownType(WellKnownType.Int32);
+            MetadataType iListType = _context.SystemModule.GetType("System.Collections", "IList");
+            MetadataType iListOfTType = _context.SystemModule.GetType("System.Collections.Generic", "IList`1");
+
+            InstantiatedType iListOfIntType =
+                iListOfTType.MakeInstantiatedType(new Instantiation(new TypeDesc[] { intType }));
+            TypeDesc intSzArrayType = intType.MakeArrayType();
+            TypeDesc intArrayType = intType.MakeArrayType(1);
+
+            Assert.True(intSzArrayType.CanCastTo(iListOfIntType));
+            Assert.True(intSzArrayType.CanCastTo(iListType));
+
+            Assert.False(intArrayType.CanCastTo(iListOfIntType));
+            Assert.True(intArrayType.CanCastTo(iListType));
+        }
+
+        [Fact]
+        public void TestArrayCasting()
+        {
+            TypeDesc intType = _context.GetWellKnownType(WellKnownType.Int32);
```

Comment:
Shouldn't you add some test for casting between array of int to array of uint and vice-versa?